staging/src/k8s.io/apiserver/pkg/server/config.go

```diff
@@ -1123,7 +1123,8 @@ func installAPI(name string, s *GenericAPIServer, c *Config) {
 	routes.Version{Version: c.EffectiveVersion.Info()}.Install(s.Handler.GoRestfulContainer)
 
 	if c.EnableDiscovery {
-		wrapped := discoveryendpoint.WrapAggregatedDiscoveryToHandler(s.DiscoveryGroupManager, s.AggregatedDiscoveryGroupManager)
+		// No peer-to-peer discovery for generic apiservers.
```

Comment:
pass in s.PeerAggregatedDiscoveryManager so generic servers *can* provide a handler in their config if they want

in WrappedHandler#ServeHTTP, use aggHandler if peerAggHandler is not provided

```diff
@@ -49,9 +50,11 @@ func (wrapped *WrappedHandler) ServeHTTP(resp http.ResponseWriter, req *http.Req
 				wrapped.aggHandler.ServeHTTP(resp, req)
 				return
 			}
-			// Serve peer-aggregated discovery by default.
-			wrapped.peerAggHandler.ServeHTTP(resp, req)
-			return
+			// Serve peer-aggregated discovery by default if a peer handler is available.
+			if wrapped.peerAggHandler != nil {
+				wrapped.peerAggHandler.ServeHTTP(resp, req)
+				return
+			}
 		}
 		wrapped.aggHandler.ServeHTTP(resp, req)
 		return

```